### ISSUE TITLE
importc types: CT sizeof, `$`, {.sizeof.}, NIM_STATIC_ASSERT

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#205]
## proposal 1: deprecate incompleteStruct
deprecate `{.incompleteStruct.}` (introduced in 2012), it's the wrong default, it's not observed (AFAIK) as evidenced by face only few structs are annotated with it. It's also un-safe as a lack of annotation would mean a struct is complete, which is often wrong.

## proposal 2: introduce {.sizeof.} for importc types
* introduce `{.sizeof.}` pragma for importc types, with following meaning:
```nim
type Foo1 {.importc.} = object
  x1: cint
const s1 = sizeof(Foo1) # CT error: Foo1 is importc and either incomplete or needs to be annotated with sizeof

type Foo2 {.importc, sizeof.} = object
  x1: cint # all fields are specicied
  c2: cstring
const s2 = sizeof(Foo2) # works

type Foo3 {.importc, sizeof: s2 .} = object
  x1: cint  # some fields may be missing but sizeof is specified
static: doAssert sizeof(Foo3) == s2
```

* Note: there should be zero additional code to compute CT sizeof for importc types, it's just the regular CT sizeof except that we remove a CT error depending on presence of `{.sizeof.}`

* we add a CT check in codegen to check for errors (eg if some header is updated and fields are added), using `NIM_STATIC_ASSERT` C/C++ macro introduced in https://github.com/nim-lang/Nim/pull/13798:
codegen inserts `NIM_STATIC_ASSERT(sizeof(Foo1) == computedSizeofFoo1)` right after the header where it's define (or right after its declaration if no header).

## proposal 3: {.importc.} for ref/ptr object types
introduce `type Foo {.importc.} = ref object`, with the meaning that the `importc` applies to the object, not the ref, to avoid the common boilerplate:
```nim
type TFoo {.importc.} = object
  x1: cint
  c2: cstring
type Foo = ref TFoo
```
where `TFoo` doesn't need to be declared.
ditto with `ptr`.

I don't think there's any use case otherwise for `{.importc.}` applying to the `ref` or `ptr` directly, since a pointer is just a pointer.

## proposal 4: `{.importc:"?".}` for importc fields that shouldn't be referenced by name in codegen
this is an alternative approach to https://github.com/nim-lang/Nim/pull/13783 for cases where some field names (eg `pad`, `reserved` etc) are un-reliable because they're OS-specific, and not part of stable ABI. eg:
```
# /usr/include/x86_64-linux-gnu/bits/stat.h
struct stat
  {
...
#ifdef __x86_64__
    __syscall_slong_t __glibc_reserved[3];
#else
# ifndef __USE_FILE_OFFSET64
    unsigned long int __glibc_reserved4;
    unsigned long int __glibc_reserved5;
# else
    __ino64_t st_ino;                   /* File serial number.  */
# endif
#endif
  };
```

```nim
type stat {.importc: "stat", header = "<stat.h>".} =
  foo1: cint
  reserved {importc:"?"}: array[4, byte]
  foo2: cstring
```

**Comment by krux02** [comment]
You propose 5 things to solve what problem?

**Comment by timotheecour** [comment]
updated top section with a summary

**Comment by cooldome** [comment]
Proposal 1: we already have pragma `size` which does this. Possibly you want to extend this pragma coverage.
Proposal 3: -1 from me, imo one line of boiler plate is better than one more convention to remember.

**Comment by krux02** [comment]
The size of an object is pretty much worthless for sizeof computations, if you don't also know the alignment.

**Comment by timotheecour** [comment]
> Proposal 6:

how is that different from proposal 4 besides the name change `{importc:"?"}` vs `{.hidden.}` ? I don't care which of these 2 pragmas is used though, `{.hidden.}` is fine with me

**Comment by krux02** [comment]
well currently `sizeof` computation does opt out of all types that are imported via `{.importc.}`. If you want to change that, this must be part of the RFC.

**Comment by timotheecour** [comment]
/cc @cooldome
> Proposal 1: we already have pragma size which does this. Possibly you want to extend this pragma coverage.

indeed, `{.size:4.}` is similar (and allows sizeof at CT), but the big difference though is to allow sizeof with 0 argument: `{.importc, sizeof.}` which implies to use `foldSizeOf`. 

Also, `.size:N` has weird semantics conflating sizeof and alignof (only `1,2,4,8` is accepted, and it implies `alignment = N` which is wrong eg for `type M1 {.importc, size:4.} = object: x: array[4, bool]`; alignment should be 1, not 4. This RFC could possibly extend `size` with this new meaning instead of introducing sizeof, TBD.

/cc @krux02 
> well currently sizeof computation does opt out of all types that are imported via {.importc.}, you should mention that you want to change that.

That's basically what proposal 1 is about; allow `.importc` types to use the `foldAlignOf`, `foldSizeOf` when `.size` or `.size:N` is specified.

> The size of an object is pretty much worthless for sizeof computations, if you don't also know the alignment.

same exact same thing would apply with `{.align.}` (infer via `foldAlignOf`)  and `{.align:N.}` (use N)
All these would be statically enforced via `NIM_STATIC_ASSERT` at cgen compile time.

> No. If those fields should be inspectable, give them the correct name of the C declaration.

you can't for ABI compatibility reasons. But you should still be able to access the values (implicitly casted to specified type) transparently: the cgen inserts `(FieldTypeInNimDecl)(offsetof(CStructType, fieldOffset))` every time a hidden field is accessed/set. So it should work transparently. Maybe a POC can show this.

> The sink _ identifier would work for this. 
`_` would make it hard to serialize/inspect the type, whereas `reserved {.hidden.}: array[N, byte]` would allow exactly that.

> You must add static asserts for all field offsets as well and the object alignment as well

yes, and it's cheap to do.

>> type Foo3 {.importc, sizeof: s2 .} = object
> That part is underspecified. What is s2? It doesn't exist in your code example. Also, what use case does it cover? After I don't think it should be supported.

s2 is any expression that can be computed at CT and is of type `int`.
Example:
```nim
const N = when defined(osx): 40 else: 48 # just an example; in practice we can automate
type cvMat {.importc, sizeof: N, alignof: cint.alignof.} = object
  rows*, cols*: cint
  data*: pointer
  # other fields that are skipped for various reasons (eg not needed, laziness, or ABI issues etc)
```
Now cvMat.{sizeof,alignof} can be used at CT, which in turns mean any type involving it can be used at CT.
Which means more code works at CT.

**Comment by krux02** [comment]
> _ would make it hard/impossible to serialize/inspect/debug the type, whereas reserved {.hidden.}: array[N, byte] would allow exactly that.

No, it would not. All fields with their true names will be visualized correctly in gdb and other debuggers, they use the debug information that comes from the C(++) compiler, with correct field names.

**Comment by timotheecour** [comment]
> _ would make it hard/impossible to serialize/inspect/debug the type
> No, it would not. All fields with their true names will be visualized correctly in gdb and other debuggers

I'm talking about from getting/setting those from nim code, not from a debugger. of course a debugger has access to those with debug info

**Comment by krux02** [comment]
> I'm talking about getting/setting those fields from nim code (eg: serialization), not from a debugger. Of course a debugger has access to those with debug info.

Give them the correct name of the C declaration, then you can access those members from Nim properly.

**Comment by timotheecour** [comment]
see https://github.com/nim-lang/Nim/pull/13926 which implements proposal 1
